### PR TITLE
Setup CI for publishing package to nuget.org

### LINKED ISSUE #5: CI status
Currently it is showing unknown in the README for both AppVeyor and TravisCI.

* TravisCI: per discussion on [this thread](https://travis-ci.community/t/status-badge-stuck-at-unknown-build/5313/3), `Build pushed branches` toggle needs to be enabled to get the badge for branches (in this case `master`): https://travis-ci.com/package-url/packageurl-dotnet/settings

* AppVeyor: please sign in with GitHub organization account which has access to settings, so the project lights up at: https://ci.appveyor.com/project/package-url/packageurl-dotnet/branch/master

### PR DESCRIPTION
**Pull request by ThomasPiskol**
This PR creates two [GitHub actions](https://github.com/features/actions) to fix #5 and #6:

1. [build.yml](https://github.com/ThomasPiskol/packageurl-dotnet/blob/d2dd5e359f605aac41a2ac3c5a478ea5f3382861/.github/workflows/build.yml)
2. [publish.yml](https://github.com/ThomasPiskol/packageurl-dotnet/blob/d2dd5e359f605aac41a2ac3c5a478ea5f3382861/.github/workflows/publish.yml)

The build actions basically runs `dotnet restore`, `dotnet build` and `dotnet test`. If the build is running on the `master` branch and all tests are successfully passed, a [GitHub release](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository) is created in draft mode.

I'm using the [GitVersion MSBuild Task](https://gitversion.net/docs/usage/msbuild) to calculate a proper version. Currently no specific configuration is done for GitVersion.
Additionally I've added an reference to the [dotnet sourcelink](https://github.com/dotnet/sourcelink) NuGet package.

Now this draft release can [be discarded or published](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository#editing-a-release). This allows the maintainer to decide, whether a new release is created by a single PR or if multiple PRs should be released together at a later point in time. Also it's possible to edit the description for this GitHub release.

It's also possible to change this behaviour and make it completely automated.

In case a release is published, the second GitHub action [publish.yml](https://github.com/ThomasPiskol/packageurl-dotnet/blob/d2dd5e359f605aac41a2ac3c5a478ea5f3382861/.github/workflows/publish.yml) is started automatically. It will download the nuget package from the GitHub release and publish it to nuget.org.

**Open Issue**: In order to publish the package to nuget.org, an API key is required. This API key can be stored as a secret in this GitHub project. It makes sense to create a [NuGet Organization](https://docs.microsoft.com/en-us/nuget/nuget-org/organizations-on-nuget-org) and use a API key generated from this organization. I can also support you in setting up this NuGet connection.

### CODE REVIEW

**Comment by JamieMagee** [comment]
Is it possible to get this reviewed?

Also, is it now a prerequisite to move CI to GitHub actions like other repos in the `package-url` org? I am happy to contribute if the maintainers will review.

Tagging @sschuberth for visibility, as you were the last person maintaining this repo.

**Comment by pombredanne** [comment]
@ThomasPiskol 

@JamieMagee suggested this on Gitter: (I invited you there):

> Hi all,
Is it possible to get a review of this PR? package-url/packageurl-dotnet#7
It also looks like the .NET purl package is lacking maintainership. I think Thomas Piskol, who opened that PR, would be a good candidate if you would like more active maintenance of that package.


 @ThomasPiskol @am11 ^ ... what do you guys want to do? @sschuberth you also helped in the past

**Comment by pombredanne** [comment]
@ThomasPiskol Do you want to take over the maintenance of this package?

**Comment by sschuberth** [comment]
> Tagging @sschuberth for visibility, as you were the last person maintaining this repo.

I really wasn't maintaining it, I just trivially updated the README at some point 😉

> @sschuberth you also helped in the past

I'm not a .NET guy, so I just skimmed over the formal aspects of this PR. As someone why enjoys reviewing PRs commit-by-commit, paying special attention to the commit messages, I miss rationales for some of the commits.

Just as an example, the "Ensure that git history is available" commit does not explain *why* that is important (I guess for `GitVersion` to be able to find a recent tag in order to describe the Git commit).

Similar for other changes, I can see from the diff *what's* being done, but the *why* is often missing, which makes it unnecessary hard for a reviewer to follow.

**Comment by ThomasPiskol** [comment]
@pombredanne Yes, I can take over the maintenance of this package.

**Comment by am11** [comment]
TBH, packageurl is a simple C# project, which does not need this amount of ceremony just to publish on git-tag-push (changes in csproj, the whole GitVersion thing).

@pombredanne, thanks for asking. I can do simple version if you prefer.

**Comment by ThomasPiskol** [comment]
Ok, so how do we move forward? 
I can explain my changes in detail, so that it's easier to follow the intention of my changes (as suggested by @sschuberth).
But this does not make sense, if we drop this PR as suggested by @am11

**Comment by pombredanne** [comment]
@ThomasPiskol @am11 I invited you both as maintainers!
You are the .net experts so I will defer to you both deciding what is best done!
Thank you ++ :bow:

**Comment by ThomasPiskol** [comment]
I've update the commit messages with an additional description, why a change was done. 
Hope this makes the review easier.

@am11: Based on my experience, it's better to automate repetitive tasks. Even if these tasks are quite simple, there's a chance that one person might forget to perform a single step in the release process. By defining these steps in code, we can eliminate this issue. Although we might get other issues 🙂

Also the entire CI description is around 100 lines of code, so it's only a small ceremony 😉

File: .github/workflows/publish.yml
Comment (coderpatros) [inline .github/workflows/publish.yml:28]: I might be mistaken, but don't you need to also publish the snupkg file as an explicit step?

**Comment by am11** [comment]
@ThomasPiskol, thanks for the update. I'm all for automation without trading simplicity. I was thinking about a self-contained approach to create GitHub Releases as well as NuGet package. The well-known release workflow of GitHub can be used to accomplish that for any language platform, and we can use `hub` utility (available in GitHub Actions by default) to create GH releases:

```yml
on:
  push:
    branches: [ master ]
    tags: '*'
  pull_request:
    branches: [ master ]

jobs:
  build:
    runs-on: ubuntu-latest
...
    steps:
...

      # only run these on tag push:

      - name: Build artifacts
        if: ${{ startsWith(github.ref, 'refs/tags/') }}
        run: dotnet pack -c Release -o package -p:Version="${GITHUB_REF#refs/tags/};CommitSha=${GITHUB_SHA}"

      - name: Create release and upload artifacts
        if: ${{ startsWith(github.ref, 'refs/tags/') }}
        run: |
          for file in package/* do
            artifacts="$artifact -a $file"
          done

          semverTag="${GITHUB_REF#refs/tags/}"
          hub release create -a publish/filename-$semverTag.nuget -m "$semverTag" "$semverTag"

          dotnet nuget push package/PackageUrl.$semverTag.nuget 
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
in the csproj:
```csproj
  <PropertyGroup>
    <Version Condition="'$(Version)' = ''">0.0.0-dev</Version>
    <CommitSha Condition="'$(CommitSha)' = ''">N/A</CommitSha>
  </PropertyGroup>
```
when it's time for a release, maintainer will do:
```sh
$ git tag 2.0.1
$ git push --tags
```
In a complex project, we may need to consider broader workflows and use targets to set commit hash etc. However, this is a simple project and we can capture source control information in release artifacts with built-in mechanics of .NET SDK and GitHub Actions; without resorting to external dependencies.

**Comment by ThomasPiskol** [comment]
@am11 thanks for your explanation 🙇 

My intention for using GitVersion is that I prefer to keep the CI and local builds symmetric. For the same commit I get basically the same result incl. the version no matter if it was built by the CI pipeline or locally on my dev machine. If GitVersion finds a valid semantic version tag on the currently checked-out commit, [this tag will be used](https://gitversion.net/docs/reference/version-sources):

> The logic of GitVersion is something like this:
> 
> - Is the current commit tagged
>   - Yes: Use the tag as the version
>   - No: continue
> - Calculate the base version (highest version from all the sources)
> - Increment version if needed based on branch config
> - Calculate the build metadata (everything after the +) and append to the calculated version
> 

For setting the Commit-Hash I would still follow the [Source Link](https://github.com/dotnet/sourcelink) approach by referencing [GitHub Source Link NuGet package](https://www.nuget.org/packages/Microsoft.SourceLink.GitHub). 
Enabling the MSBuild property `ContinuousIntegrationBuild` is required, because this guarantees [deterministic builds](https://devblogs.microsoft.com/dotnet/producing-packages-with-source-link/#deterministic-builds) by normalizing the source paths. 

Btw, the Source Link package and the GitVersion package are just development dependencies and are not required for consumers of the packageurl-dotnet NuGet package. That's because of setting `PrivateAssets` to `all` in the **PackageUrl.csproj**.

My proposal is:
1. Change the workflow, so that a release is automatically created in GitHub and the NuGet package is automatically uploaded to nuget.org if a maintainer pushes a new tag.
2. Drop the usage of `softprops/action-gh-release` and directly use the GitHub CLI tool for creating the release.
3. Keep the Source Link reference because it's the "official" .NET way.
4. Keep GitVersion to have a "symmetric" build.

What do you think?

**Comment by dfederm** [comment]
@ThomasPiskol I like your approach to versioning and it's what we use in more or less all our projects. Although we use NerdBank.GitVersioning instead of GitVersion, but it sounds very similar in concept. @am11 Any concerns?

**Comment by dfederm** [comment]
@ThomasPiskol  Gentle ping :)

**Comment by ThomasPiskol** [comment]
Sorry, due to unexpected events I was not able to work on the PR in the past weeks. 
To integrate this PR it needs to be approved by someone else (I'm not able to approve my own PR).
@am11, @sschuberth can you review the latest changes?

**Comment by ThomasPiskol** [comment]
@am11 @sschuberth @pombredanne 
How can we move forward with this PR?

**Review by am11 — APPROVED** [review]
Thanks you! And sorry, I missed the notifications and it fell off my radar.
Changes looks good to me. 👍

**Pull request merged**
